src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf8StringMarshaller.cs

```diff
@@ -80,15 +83,19 @@ public Utf8StringMarshaller(string? str, Span<byte> buffer)
         /// <remarks>
         /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
         /// </remarks>
-        public void FromNativeValue(byte* value) => _allocated = value;
+        public void FromNativeValue(byte* value)
+        {
+            _nativeValue = value;
+            _allocated = true;
+        }
 
         /// <summary>
         /// Returns the managed string.
         /// </summary>
         /// <remarks>
         /// <seealso cref="CustomTypeMarshallerDirection.Out"/>
         /// </remarks>
-        public string? ToManaged() => _allocated == null ? null : Marshal.PtrToStringUTF8((IntPtr)_allocated);
+        public string? ToManaged() => Marshal.PtrToStringUTF8((IntPtr)_nativeValue);
```

Comment:
```suggestion
        public string? ToManaged() => ConvertToManaged(_nativeValue);
```

Reply:
I have somehow mixed up the draft change that is refactoring the built-in marshallers into this PR.  

I am going to remove the commit that is refactoring the built-in marshallers. It needs some more work.